controllers/machineset_controller.go

```diff
@@ -126,21 +127,41 @@ func (r *ReconcileMachineSet) ProcessMachineSet(ctx context.Context, machineSet
 		// Update labels in machine
 		err = r.updateLabelsInMachine(ctx, machine, expectedLabels)
 		if err != nil {
+			metrics.RegisterFailedLabelUpdate(&metrics.MnmoLabels{
+				Name:      machine.Name,
+				Namespace: machine.Namespace,
+				Reason:    "Machine label update failed",
+			})
 			return reconcile.Result{}, err
 		}
 		// Update taints in machine
 		err = r.updateTaintsInMachine(ctx, machineSet, machine)
 		if err != nil {
+			metrics.RegisterFailedTaintUpdate(&metrics.MnmoLabels{
+				Name:      machine.Name,
+				Namespace: machine.Namespace,
+				Reason:    "Machine taint update failed",
+			})
 			return reconcile.Result{}, err
 		}
 		//Update labels in node
 		err = r.updateLabelsInNode(ctx, node, expectedLabels)
 		if err != nil {
+			metrics.RegisterFailedLabelUpdate(&metrics.MnmoLabels{
+				Name:      node.Name,
+				Namespace: node.Namespace,
+				Reason:    "Node label update failed",
+			})
 			return reconcile.Result{}, err
 		}
 		// Update taints in node
 		err = r.updateTaintsInNode(ctx, machine, node)
 		if err != nil {
+			metrics.RegisterFailedTaintUpdate(&metrics.MnmoLabels{
+				Name:      node.Name,
+				Namespace: node.Namespace,
+				Reason:    "Node taint update failed",
```

Comment:
I would simply remove the `Reason` from the labels.
Firstly, the `Reason` mentioned here isn't a reason, but the sympton and just mirrors what's already in the metric name and secondly, monitoring should focus on symptoms and the state of the system. Finding out the reason for certain failures should be done using logs. Thirdy, if we decide to go forward with this label, the cardinality of the metric is pretty much unbound.